src/Disks/IO/ThreadPoolReader.cpp

```diff
@@ -200,9 +200,9 @@ std::future<IAsynchronousReader::Result> ThreadPoolReader::submit(Request reques
 
     ProfileEvents::increment(ProfileEvents::ThreadPoolReaderPageCacheMiss);
 
-    ThreadGroupStatusPtr running_group = CurrentThread::isInitialized() && CurrentThread::get().getThreadGroup()
-            ? CurrentThread::get().getThreadGroup()
-            : MainThreadStatus::getInstance().getThreadGroup();
+    ThreadGroupStatusPtr running_group;
+    if (CurrentThread::isInitialized() && CurrentThread::get().getThreadGroup())
```

Comment:
I am not familiar with this place. Why do not we use deleter below ?
```
if (query_context)
    thread_status.attachQueryContext(query_context);
if (running_group)
    thread_status.attachQuery(running_group);
```
If exception will be thrown somewhere between attach and detach will we leak something ?
```
if (running_group)
    thread_status.detachQuery();
```

Reply:
>If exception will be thrown somewhere between attach and detach will we leak something ?

You are right.

>I am not familiar with this place. Why do not we use deleter below ?

I will use `SCOPE_EXIT` here instead.

*`deleter` is not a common thing to use in a code base, however it is required in `buildPushingToViewsChain()` because it should capture some pointers*